api/client/set.go

```diff
@@ -0,0 +1,67 @@
+package client
+
+import (
+	"fmt"
+
+	flag "github.com/docker/docker/pkg/mflag"
+	"github.com/docker/docker/pkg/units"
+	"github.com/docker/docker/runconfig"
+)
+
+func (cli *DockerCli) CmdSet(args ...string) error {
+	cmd := cli.Subcmd("set", "CONTAINER [CONTAINER...]", "Setup/Update resource configs for a container", true)
+	flCpusetCpus := cmd.String([]string{"#-cpuset", "-cpuset-cpus"}, "", "CPUs in which to allow execution (0-3, 0,1)")
+	flCpusetMems := cmd.String([]string{"-cpuset-mems"}, "", "MEMs in which to allow execution (0-3, 0,1)")
+	flCpuShares := cmd.Int64([]string{"c", "-cpu-shares"}, 0, "CPU shares (relative weight)")
+	flMemoryString := cmd.String([]string{"m", "-memory"}, "", "Memory limit")
+	flMemorySwap := cmd.String([]string{"-memory-swap"}, "", "Total memory (memory + swap), '-1' to disable swap")
+	flCpuQuota := cmd.Int64([]string{"-cpu-quota"}, 0, "Limit the CPU CFS (Completely Fair Scheduler) quota")
+	cmd.Require(flag.Min, 1)
+
+	cmd.ParseFlags(args, true)
+
+	var flMemory int64
+	if *flMemoryString != "" {
+		parsedMemory, err := units.RAMInBytes(*flMemoryString)
+		if err != nil {
+			return err
+		}
+		flMemory = parsedMemory
+	}
+
+	var MemorySwap int64
+	if *flMemorySwap != "" {
+		if *flMemorySwap == "-1" {
+			MemorySwap = -1
+		} else {
+			parsedMemorySwap, err := units.RAMInBytes(*flMemorySwap)
+			if err != nil {
+				return err
+			}
+			MemorySwap = parsedMemorySwap
+		}
+	}
+
+	hostConfig := runconfig.HostConfig{
```

Comment:
I don't think this is specific to this PR because I've been thinking about it for while but haven't found a situation where it actually manifested itself as an issue for an end-user until now, but the way we parse configs and then try to merge config structs is a problem.

If I run: `docker set -c 0 0685bc59c0dadfbb9971` the daemon has no way of knowing if zero means "the user specified zero" or "the user didn't specify it at all".  We really need to have the notion of sending a structure that has **just** the changed fields, no default values - then we know exactly what the user specified.

I would recommend that we:
- just send a json object with just the modified fields, no defaults
- have the daemon process each field individually instead of trying to deal with it as a single hostConfig struct